### PR TITLE
convert repository to use gulp patterns (closes #357)

### LINKED ISSUE #357: es6 and compile to es5, eslint
Tired of jshint and I am used to es6 enough of this legacy stuff.

### PR DESCRIPTION
**Pull request by phated**
Since it doesn't seem like people know how to use the gulp patterns to have eslint/coveralls/etc, I've started working on converting this repo over to those patterns and continue to support node 0.10+

### CODE REVIEW

**Comment by nmccready** [comment]
Looks good so far!

**Comment by phated** [comment]
Only the 695 line `write.js` file remains!  I should be able to wrap that up tomorrow.

**Comment by phated** [comment]
@nmccready this should be all set!  I'll take another look at everything and make sure nothing was missed.  Once we are ready to merge it, we'll probably want to squash it into 1 commit.

**Comment by phated** [comment]
@nmccready I'm totally to changing rules in https://github.com/gulpjs/eslint-config-gulp if you'd like to make changes - I did make a few local changes to make warnings go away.

**Comment by phated** [comment]
After this PR, I plan to enable AppVeyor CI and fix any Windows issues I find, but I didn't want to combine them in this one/squash that into this commit.

File: .eslintrc [hunk 1]
```diff
@@ -1,3 +1,9 @@
 {
-  "extends": "gulp"
+  "extends": "gulp",
+  "rules": {
+    "max-depth": [1, 5],
```
Comment (nmccready) [inline .eslintrc]: 👍

File: .gitignore [hunk 2]
```diff
@@ -1,9 +1,31 @@
...
+build/Release
+
+# Dependency directory
+# Commenting this out is preferred by some people, see
+# https://www.npmjs.org/doc/misc/npm-faq.html#should-i-check-my-node_modules-folder-into-git-
 node_modules
-reports
-*.tgz
-.idea
-*.lock
```
Comment (nmccready) [inline .gitignore]: Are we allowing lock files now? Or is this covered with `.lock-wscript`?

File: .travis.yml [hunk 3]
```diff
@@ -1,23 +1,11 @@
 sudo: false
-
 language: node_js
-
 node_js:
+  - '10'
+  - '8'
   - '6'
   - '4'
   - '0.12'
```
Comment (nmccready) [inline .travis.yml]: When will abandoning 0.12 and 0.10 be on our radar?

File: test/utils.js [hunk 4]
```diff
@@ -1,31 +1,34 @@
 'use strict';
 
-var test = require('tape');
+var expect = require('expect');
 var utils = require('../src/utils');
 
-test('exceptionToString: takes message if present', function(t) {
-  var exception = { message: 'exception message' };
-  var result = utils.exceptionToString(exception);
+describe('utils', function() {
```
Comment (nmccready) [inline test/utils.js]: Besides switching to mocha, how do you feel about making all the acual test files be `*.test.js` or `*.spec.js` ?

**Review by nmccready — CHANGES_REQUESTED** [review]
Looks great to me, just a few comments to have a discussion.

File: .gitignore (hunk 2 above)
Comment (phated) [reply .gitignore]: good catch

File: .travis.yml (hunk 3 above)
Comment (phated) [reply .travis.yml]: I'm planning to drop it in gulp 5

File: test/utils.js (hunk 4 above)
Comment (phated) [reply test/utils.js]: Done

**Comment by phated** [comment]
@nmccready addressed your comments.  Hopefully we are good to merge this.

File: coverage/coverage.json [hunk 5]
```diff
@@ -1 +0,0 @@
...:39}},{"start":{"line":127,"column":44},"end":{"line":127,"column":64}},{"start":{"line":127,"column":68},"end":{"line":127,"column":99}}]},"20":{"line":133,"type":"binary-expr","locations":[{"start":{"line":133,"column":33},"end":{"line":133,"column":46}},{"start":{"line":133,"column":50},"end":{"line":133,"column":78}}]},"21":{"line":149,"type":"if","locations":[{"start":{"line":149,"column":6},"end":{"line":149,"column":6}},{"start":{"line":149,"column":6},"end":{"line":149,"column":6}}]},"22":{"line":151,"type":"if","locations":[{"start":{"line":151,"column":8},"end":{"line":151,"column":8}},{"start":{"line":151,"column":8},"end":{"line":151,"column":8}}]},"23":{"line":161,"type":"if","locations":[{"start":{"line":161,"column":6},"end":{"line":161,"column":6}},{"start":{"line":161,"column":6},"end":{"line":161,"column":6}}]},"24":{"line":161,"type":"binary-expr","locations":[{"start":{"line":161,"column":10},"end":{"line":161,"column":34}},{"start":{"line":161,"column":38},"end":{"line":161,"column":84}}]},"25":{"line":168,"type":"if","locations":[{"start":{"line":168,"column":4},"end":{"line":168,"column":4}},{"start":{"line":168,"column":4},"end":{"line":168,"column":4}}]}}}}
```
Comment (nmccready) [inline coverage/coverage.json]: Hah my goof

**Review by nmccready — APPROVED** [review]

File: src/write/index.internals.js [hunk 6]
```diff
@@ -32,34 +32,34 @@ module.exports = function(destPath, options) {
...
+      debug(function() { return 'filePath: ' + filePath; });
 
-      if (options.mapSourcesAbsolute === true){
+      if (options.mapSourcesAbsolute === true) {
         debug(function() { return 'mapSourcesAbsolute'; });
 
-        if (!file.dirname){
+        if (!file.dirname) {
           debug(function() { return '!file.dirname'; });
           filePath = path.join(file.base, filePath).replace(file.cwd, '');
```
Comment (coreyfarrell) [inline src/write/index.internals.js]: Before this PR the tests hit this line, now it does not as `file.dirname` is never false.

Comment (phated) [reply src/write/index.internals.js]: I believe that's due to the vinyl or gulp 4 update. I think it's fine to leave for older vinyl or gulp.

**Review by coreyfarrell — APPROVED** [review]

**Comment by phated** [comment]
:shipit:

**Pull request merged**